File: src/types/project.ts
export type TableId = string;
export type AnalysisId = string;
export type GraphId = string;
export type LayoutId = string;

export type TableFormatId =
  | 'xy'
  | 'column'
  | 'grouped'
  | 'contingency'
  | 'survival'
  | 'partsOfWhole'
  | 'multipleVariables'
  | 'nested';

export type AnalysisTypeId =
  | 'descriptive'
  | 'unpaired_ttest'
  | 'paired_ttest'
  | 'one_way_anova'
  | 'two_way_anova'
  | 'chi_square'
  | 'fisher_exact'
  | 'kaplan_meier'
  | 'fraction_of_total'
  | 'mann_whitney'
  | 'kruskal_wallis'
  | 'roc_auc'
  | 'normality_test'
  | 'linear_regression'
  | 'dose_response_4pl';

export type GraphTypeId =
  | 'bar'
  | 'groupedBar'
  | 'scatter'
  | 'line'
  | 'scatterLine'
  | 'survival'
  | 'doseResponse'
  | 'pie';

export interface ColumnTableData {
  columnLabels: string[];
  rows: (number | null)[][];
  /** When set, columns are replicates within groups; groupForColumn[i] = index into groupLabels */
  groupLabels?: string[];
  groupForColumn?: number[];
}

export interface XYTableData {
  xLabel: string;
  yLabels: string[];
  x: (number | null)[];
  ys: (number | null)[][];
}

/** Two grouping variables; each cell holds replicate values. @spec PRISM-TBL-004 */
export interface GroupedTableData {
  rowGroupLabels: string[];
  colGroupLabels: string[];
  /** cellValues[rowIdx][colIdx] = replicate values for that cell */
  cellValues: (number | null)[][][];
}

/** Integer counts per row/column category. @spec PRISM-TBL-005 */
export interface ContingencyTableData {
  rowLabels: string[];
  columnLabels: string[];
  counts: number[][];
}

/** Time-to-event; one row per subject. @spec PRISM-TBL-006 */
export interface SurvivalTableData {
  timeLabel: string;
  eventLabel: string;
  groupLabel?: string;
  times: number[];
  events: number[];
  groups?: string[];
}

/** Labels + values (e.g. fractions). @spec PRISM-TBL-007 */
export interface PartsOfWholeTableData {
  labels: string[];
  values: number[];
}

export interface DataTable {
  id: TableId;
  name: string;
  format: TableFormatId;
  data:
    | ColumnTableData
    | XYTableData
    | GroupedTableData
    | ContingencyTableData
    | SurvivalTableData
    | PartsOfWholeTableData;
}

export const CURRENT_PROJECT_VERSION = 1;

/** @spec PRISM-WKF-006 Layout item: one graph placed in the layout. */
export interface LayoutItem {
  graphId: GraphId;
  /** Position and size in 0–1 fraction of layout (x, y, width, height). */
  x: number;
  y: number;
  width: number;
  height: number;
}

/** @spec PRISM-WKF-006 Layout sheet composing multiple graphs. */
export interface Layout {
  id: LayoutId;
  name: string;
  items: LayoutItem[];
}

export type Selection =
  | { type: 'table'; tableId: TableId }
  | { type: 'analysis'; analysisId: AnalysisId }
  | { type: 'graph'; graphId: GraphId }
  | { type: 'layout'; layoutId: LayoutId }
  | null;

import type { Analysis } from './analysis';
import type { Graph } from './graph';

export interface Project {
  version: number;
  tables: DataTable[];
  analyses: Analysis[];
  graphs: Graph[];
  layouts: Layout[];
  selection: Selection;
}

File: backend/app/models/project.py
"""Data table types and project structure for Dataplotter."""

from __future__ import annotations

from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, Field


class TableFormatId(str, Enum):
    XY = "xy"
    COLUMN = "column"
    GROUPED = "grouped"
    CONTINGENCY = "contingency"
    SURVIVAL = "survival"
    PARTS_OF_WHOLE = "partsOfWhole"
    MULTIPLE_VARIABLES = "multipleVariables"
    NESTED = "nested"


class AnalysisTypeId(str, Enum):
    DESCRIPTIVE = "descriptive"
    UNPAIRED_TTEST = "unpaired_ttest"
    PAIRED_TTEST = "paired_ttest"
    ONE_WAY_ANOVA = "one_way_anova"
    TWO_WAY_ANOVA = "two_way_anova"
    CHI_SQUARE = "chi_square"
    FISHER_EXACT = "fisher_exact"
    KAPLAN_MEIER = "kaplan_meier"
    FRACTION_OF_TOTAL = "fraction_of_total"
    MANN_WHITNEY = "mann_whitney"
    KRUSKAL_WALLIS = "kruskal_wallis"
    ROC_AUC = "roc_auc"
    NORMALITY_TEST = "normality_test"
    LINEAR_REGRESSION = "linear_regression"
    DOSE_RESPONSE_4PL = "dose_response_4pl"
    CORRELATION = "correlation"
    MULTIPLE_REGRESSION = "multiple_regression"
    NESTED_TTEST = "nested_ttest"
    NESTED_ONE_WAY_ANOVA = "nested_one_way_anova"


class GraphTypeId(str, Enum):
    BAR = "bar"
    BOX = "box"
    GROUPED_BAR = "groupedBar"
    SCATTER = "scatter"
    LINE = "line"
    SCATTER_LINE = "scatterLine"
    SURVIVAL = "survival"
    DOSE_RESPONSE = "doseResponse"
    PIE = "pie"
    PARAMETER_BAR = "parameterBar"


TableId = Annotated[str, Field(pattern=r"^[a-zA-Z0-9_-]+$")]
AnalysisId = Annotated[str, Field(pattern=r"^[a-zA-Z0-9_-]+$")]
GraphId = Annotated[str, Field(pattern=r"^[a-zA-Z0-9_-]+$")]
LayoutId = Annotated[str, Field(pattern=r"^[a-zA-Z0-9_-]+$")]


class ColumnTableData(BaseModel):
    column_labels: list[str] = Field(..., alias="columnLabels")
    rows: list[list[float | None]]
    group_labels: list[str] | None = Field(None, alias="groupLabels")
    group_for_column: list[int] | None = Field(None, alias="groupForColumn")

    class Config:
        populate_by_name = True


class XYTableData(BaseModel):
    x_label: str = Field(..., alias="xLabel")
    y_labels: list[str] = Field(..., alias="yLabels")
    x: list[float | None]
    ys: list[list[float | None]]

    class Config:
        populate_by_name = True


class GroupedTableData(BaseModel):
    row_group_labels: list[str] = Field(..., alias="rowGroupLabels")
    col_group_labels: list[str] = Field(..., alias="colGroupLabels")
    cell_values: list[list[list[float | None]]] = Field(..., alias="cellValues")

    class Config:
        populate_by_name = True


class ContingencyTableData(BaseModel):
    row_labels: list[str] = Field(..., alias="rowLabels")
    column_labels: list[str] = Field(..., alias="columnLabels")
    counts: list[list[int]]

    class Config:
        populate_by_name = True


class SurvivalTableData(BaseModel):
    time_label: str = Field(..., alias="timeLabel")
    event_label: str = Field(..., alias="eventLabel")
    subject_labels: list[str] | None = Field(None, alias="subjectLabels")
    group_labels: list[str] | None = Field(None, alias="groupLabels")
    times: list[float]
    events: list[int]
    groups: list[str] | None = None

    class Config:
        populate_by_name = True


class PartsOfWholeTableData(BaseModel):
    labels: list[str]
    values: list[float]

    class Config:
        populate_by_name = True


class MultipleVariablesTableData(BaseModel):
    variable_labels: list[str] = Field(..., alias="variableLabels")
    rows: list[list[float | None]]

    class Config:
        populate_by_name = True


class NestedTableData(BaseModel):
    column_labels: list[str] = Field(..., alias="columnLabels")
    rows: list[list[float | None]]
    group_labels: list[str] | None = Field(None, alias="groupLabels")
    group_for_column: list[int] | None = Field(None, alias="groupForColumn")

    class Config:
        populate_by_name = True


TableData = (
    ColumnTableData
    | XYTableData
    | GroupedTableData
    | ContingencyTableData
    | SurvivalTableData
    | PartsOfWholeTableData
    | MultipleVariablesTableData
    | NestedTableData
)


class TableViewMode(str, Enum):
    RAW = "raw"
    TRANSFORMED = "transformed"


class DataMode(str, Enum):
    RAW = "raw"
    TRANSFORMED = "transformed"


class ColumnTransformation(BaseModel):
    column_key: str = Field(..., alias="columnKey")
    equation: str

    class Config:
        populate_by_name = True


CURRENT_PROJECT_VERSION = 1


class DataTable(BaseModel):
    id: TableId
    name: str
    format: TableFormatId
    data: TableData
    transformations: list[ColumnTransformation] | None = None
    view_mode: TableViewMode | None = Field(None, alias="viewMode")
    table_data_version: int | None = Field(None, alias="tableDataVersion")

    class Config:
        populate_by_name = True


class LayoutItem(BaseModel):
    graph_id: GraphId = Field(..., alias="graphId")
    x: float
    y: float
    width: float
    height: float

    class Config:
        populate_by_name = True


class Layout(BaseModel):
    id: LayoutId
    name: str
    items: list[LayoutItem]


class TableSelection(BaseModel):
    type: Literal["table"] = "table"
    table_id: TableId = Field(..., alias="tableId")


class AnalysisSelection(BaseModel):
    type: Literal["analysis"] = "analysis"
    analysis_id: AnalysisId = Field(..., alias="analysisId")


class GraphSelection(BaseModel):
    type: Literal["graph"] = "graph"
    graph_id: GraphId = Field(..., alias="graphId")


class LayoutSelection(BaseModel):
    type: Literal["layout"] = "layout"
    layout_id: LayoutId = Field(..., alias="layoutId")


SelectionType = Annotated[
    TableSelection | AnalysisSelection | GraphSelection | LayoutSelection | None,
    Field(discriminator="type"),
]


class Project(BaseModel):
    version: int = CURRENT_PROJECT_VERSION
    tables: list[DataTable]
    analyses: list["Analysis"]
    graphs: list["Graph"]
    layouts: list[Layout]
    selection: SelectionType = None


from backend.app.models.analysis import Analysis
from backend.app.models.graph import Graph

Project.model_rebuild()
DataTable.model_rebuild()
